src/main/java/com/vinylteam/vinyl/web/servlets/ConfirmationServlet.java

```diff
@@ -0,0 +1,111 @@
+        }
+        Optional<ConfirmationToken> optionalConfirmationToken = confirmationService.findByToken(token);
+        response.setContentType("text/html;charset=utf-8");
+        Map<String, String> attributes = new HashMap<>();
+        if (optionalConfirmationToken.isEmpty()) {
+            response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+            log.debug("Set response status to {'status':{}}", HttpServletResponse.SC_NOT_FOUND);
+            attributes.put("message", "Sorry, such confirmation link does not exist. Maybe your email is already verified? Try to log in! " +
+                    "If you haven't verified your email and you can't use confirmation link, please, contact us through the form.");
+            PageGenerator.getInstance().process("confirmation-directions", attributes, response.getWriter());
+        } else {
+            response.setStatus(HttpServletResponse.SC_OK);
+            log.debug("Set response status to {'status':{}}", HttpServletResponse.SC_OK);
+            attributes.put("token", tokenAsString);
+            PageGenerator.getInstance().process("confirmation-signin", attributes, response.getWriter());
+        }
+    }
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String tokenAsString = request.getParameter("token");
+        UUID token = null;
+        try {
+            token = UUID.fromString(tokenAsString);
+        } catch (IllegalArgumentException e) {
+            log.warn("Invalid email confirmation link token parameter {'token':{}}", tokenAsString, e);
+        }
+        Optional<ConfirmationToken> optionalConfirmationTokenByLinkToken = confirmationService.findByToken(token);
+        response.setContentType("text/html;charset=utf-8");
+        Map<String, String> attributes = new HashMap<>();
+        if (optionalConfirmationTokenByLinkToken.isEmpty()) {
+            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+            log.debug("Set response status to {'status':{}}", HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+            attributes.put("message", "Sorry, something went wrong on out side, we're looking into it. Please, try to login again, or contact us.");
+            PageGenerator.getInstance().process("confirmation-directions", attributes, response.getWriter());
+        } else {
+            ConfirmationToken confirmationTokenByLinkToken = optionalConfirmationTokenByLinkToken.get();
+            User userByLinkToken = userService.findById(confirmationTokenByLinkToken.getUserId()).get();
+            String email = request.getParameter("email");
+            if (userByLinkToken.getEmail().equalsIgnoreCase(email)) {
+                String password = request.getParameter("password");
+                Optional<User> optionalUser = userService.signInCheck(email, password);
+                log.debug("Received a optional with User with password verification by the passed " +
```

Comment:
Even when servlet works with two types of data?
Because I can't take confirmationToken or user out from servlet and have it working properly.
And for different entities should exist different services, right?

Reply:
One service can use another, in your case userService can use confirmationService - it is fine